### PR TITLE
Introduce SSO Session Cookie

### PR DESCRIPTION
**Pull request by Stephan-Kok**
As a Service Provider I want to be able to discover whether a user can log in or is logged in via Engineblock,
so that I can optimize the start of the authentication process.

When the user is successfully authenticated by the Identity Provider, Engineblock must store an SSO session cookie in
their browser. With this cookie it is possible, if necessary, to verify whether the user has been successfully
authenticated with at least one Identity Provider.

With this functionality it is possible for a service provider to check whether the user has a valid SSO session. If this were to be the case they can; skip the login screen, directly login the user and grant them access to the request content.

An authentication with the Identity Provider is successful when the SAML response contains:

    <samlp:Status>
        <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success" />
    </samlp:Status>

The value of the cookie contains an array of the entity ID('s) of the successfully authenticated Identity Providers.
This cookie will be deleted at `/logout` or will expire according to the set `sso_session_cookie_max_age`.
If the SSO Cookie Session feature is enabled it is also possible to manually delete this cookie at
`/authentication/idp/remove-cookies`. Do note however, that the remove cookies endpoint is only available if
`wayf.remember_choice=true`

---
### SSO query
In this PR we implemented an SSO cookie that will be stored by Engineblock after the user has been succesfully authenticated by the IdP. With SSO Query, we want to implement a standalone SSO query microservice for the OpenConext platform.

The SSO Query service will me implemented with the following logic:
- A Service Provider sends an SSO query to the SSO query service of the OpenConext platform
- The SSO query service evaluates the user's SSO session status
  - Status _TRUE_ (user has already been authenticated via OpenConext): SSO query immediately sends a response to the Service Provider
  - Status _REMOTE_ (there is an SSO notification cookie on the OpenConext domain):
  - Status _FALSE_ (there is no SSO session or SSO notification cookie on the OpenConext domain yet)

We will share the SSO Query service once it is finished.

### CODE REVIEW

**Comment by thijskh** [comment]
What is the difference between this approach and the SSO Notification feature? It seems both have a very similar functional goals.

**Comment by Stephan-Kok** [comment]
Hi Thijs,

There is a difference between the SSO Notification feature (together with SSO Notification Service) and the SSO Session feature ( together with SSO Query service).

The SSO Notifications actually **notifies** the WAYF which IDP to preselected, so it can skip the list of choices.

This enables schools to reduce the number of steps users have to take during the login process and increase user-friendliness by providing the possibility to skip the WAYF (Where Are You From) screen. This is done via the use of a cookie that is set using the SSO Notification Service, which can be set for example after the user has logged in on his own environment (for example an intranet page or the homepage of an Electronic Learning Environment).

The SSO Query actually **queries** EngineBlock to determine the status of a user.

With the SSO Query service it is possible for a Service Provider to check if the user is known and has a valid Single Sign-on Session. This SSO cookie is set by Engineblock and read from the SSO Query service. This way the Service Provider is able to request the authentication status of the user. And if the authentication is complete the users can skip the login screen all together.

We have put the topic of discussing the SSO Session / SSO Query functionality on the agenda for the next meeting.

Hope this clarifies things and if not feel free to contact me meanwhile.

Regards,
Stephan

**Comment by thijskh** [comment]
Any updates on this @Stephan-Kok?

File: app/config/config.yml [hunk 1]
```diff
@@ -25,6 +25,7 @@ open_conext_engine_block:
         eb.block_user_on_violation: "%feature_block_user_on_violation%"
         eb.enable_sso_notification: "%feature_enable_sso_notification%"
         eb.feature_enable_consent: "%feature_enable_consent%"
+        eb.enable_sso_session_cookie: "%feature_enable_sso_session_cookie%"
```
[the remark this answers is not in the document]
Comment (Stephan-Kok) [reply app/config/config.yml:28]: Yes as explained above, the SSO session is a completely different feature and they do not rely on each other in any way. More use cases are described below:

Sso Query - its a check in engineblock for an active session:
- Impove user functionality by automatically logging in an user if they have an active session (skipping the loggin button)
- Grant access to additional content based on if the user has 'a' session in EB

SSO Notification - its an action where a SSO notification cookie is placed:
- Skip WAYF screen and request an user to login with a specific IDP

**Comment by Stephan-Kok** [comment]
I have implemented the requested changes.

**Review by thijskh — APPROVED** [review]

**Pull request merged**